internal/mid/auth.go

```diff
@@ -0,0 +1,58 @@
+package mid
+
+import (
+	"context"
+	"log"
+	"net/http"
+	"strings"
+
+	"github.com/ardanlabs/service/internal/auth"
+	"github.com/ardanlabs/service/internal/platform/web"
+	"github.com/pkg/errors"
+)
+
+// Auth is used to authenticate HTTP requests.
+type Auth struct {
+	Parser *auth.Parser
+}
+
+// Authenticate validates a JWT from the `Authorization` header.
+func (a *Auth) Authenticate(next web.Handler) web.Handler {
+	h := func(ctx context.Context, log *log.Logger, w http.ResponseWriter, r *http.Request, params map[string]string) error {
+		tknStr, err := getBearerToken(r)
+		if err != nil {
+			return errors.Wrap(web.ErrUnauthorized, err.Error())
+		}
+
+		claims, err := a.Parser.ParseClaims(tknStr)
+		if err != nil {
+			return errors.Wrap(web.ErrUnauthorized, err.Error())
+		}
+
+		// TODO: Pass claims down.
+		// Options:
+		// A. Pass via context
+		// B. Pass via additional param
+		_ = claims
+
+		return next(ctx, log, w, r, params)
+	}
+
+	return h
+}
+
+// getBearerToken grabs a token from the request header. Expected header is of
+// the format `Authorization: Bearer <token>`.
+func getBearerToken(r *http.Request) (string, error) {
+	hdr := r.Header.Get("Authorization")
```

Comment:
This is where we get into fun conversations about not hiding information and testability. We could pass the "Authorization" string into this function and make it more precise. It's the only thing we need from the request. Have the calling function do this. The calling function Needs the request so it makes more sense there . IMO.

This also makes the function more testable.